File: examples/vulnerable.php
<?php
// SQL Injection vulnerabilities
function vulnerableQuery($userInput) {
    // CWE-89: SQL Injection
    $query = "SELECT * FROM users WHERE id = " . $_GET['id'];
    mysql_query($query);
    
    // Another SQL injection
    mysqli_query($connection, "SELECT * FROM products WHERE name = '" . $_POST['name'] . "'");
    
    // PDO SQL injection
    $pdo->query("SELECT * FROM orders WHERE user_id = " . $_GET['user_id']);
}

// Command Injection vulnerabilities
function vulnerableCommand() {
    // CWE-78: Command Injection
    exec($_GET['cmd']);
    system($_POST['command']);
    shell_exec($_REQUEST['shell']);
    passthru($_GET['exec']);
    
    // Code injection
    eval($_POST['code']);
}

// File Inclusion vulnerabilities
function vulnerableInclusion() {
    // CWE-98: File Inclusion
    include($_GET['page']);
    require($_POST['file']);
    include_once($_REQUEST['module']);
    require_once($_GET['lib']);
}

// Cross-Site Scripting (XSS)
function vulnerableOutput() {
    // CWE-79: XSS
    echo $_GET['message'];
    print $_POST['content'];
    printf($_REQUEST['data']);
}

// Path Traversal vulnerabilities
function vulnerableFileAccess() {
    // CWE-22: Path Traversal
    file_get_contents($_GET['file']);
    file_put_contents($_POST['filename'], "data");
    fopen($_REQUEST['path'], "r");
    readfile($_GET['document']);
}

// Weak Cryptography
function weakCrypto($data) {
    // CWE-327: Weak cryptographic algorithms
    $hash1 = md5($data);
    $hash2 = sha1($data);
    $encrypted = crypt($data);
    
    return $hash1 . $hash2 . $encrypted;
}

// Hardcoded Secrets
function hardcodedCredentials() {
    // CWE-798: Hardcoded credentials
    $password = "admin123";
    $apiKey = "sk_live_abcdef1234567890";
    $dbPassword = "SuperSecretDBPass!";
    $secretKey = "MySecretEncryptionKey2024";
    
    return array($password, $apiKey, $dbPassword, $secretKey);
}

// Insecure Random Number Generation
function weakRandom() {
    // Weak random number generation
    $token = rand();
    $sessionId = mt_rand();
    
    return $token . $sessionId;
}

// Insecure File Upload
function vulnerableUpload() {
    if (isset($_FILES['upload'])) {
        // No validation - allows any file type
        move_uploaded_file($_FILES['upload']['tmp_name'], 
                          "uploads/" . $_FILES['upload']['name']);
    }
}

// Session Fixation
function sessionFixation() {
    // Session fixation vulnerability
    session_start();
    if (isset($_GET['sessionid'])) {
        session_id($_GET['sessionid']);
    }
}

// LDAP Injection
function ldapInjection($username) {
    // CWE-90: LDAP Injection
    $filter = "(uid=" . $username . ")";
    ldap_search($connection, "dc=example,dc=com", $filter);
}

// XML External Entity (XXE)
function xxeVulnerability($xmlData) {
    // CWE-611: XXE
    $dom = new DOMDocument();
    $dom->loadXML($xmlData, LIBXML_NOENT | LIBXML_DTDLOAD);
}

// Insecure Deserialization
function insecureDeserialization() {
    // CWE-502: Insecure Deserialization
    $data = unserialize($_POST['data']);
    return $data;
}

// Information Disclosure
function informationDisclosure() {
    // Exposing sensitive information
    phpinfo();
    error_reporting(E_ALL);
    ini_set('display_errors', 1);
    
    // Database error exposure
    mysql_query("INVALID SQL") or die(mysql_error());
}

// Main execution
if ($_SERVER['REQUEST_METHOD'] === 'GET') {
    vulnerableQuery($_GET['input']);
    vulnerableCommand();
    vulnerableInclusion();
    vulnerableOutput();
    vulnerableFileAccess();
    echo weakCrypto("sensitive data");
    print_r(hardcodedCredentials());
    echo weakRandom();
    vulnerableUpload();
    sessionFixation();
    ldapInjection($_GET['user']);
    xxeVulnerability($_POST['xml']);
    var_dump(insecureDeserialization());
    informationDisclosure();
}
?>

File: examples/vulnerable.go
package main

import (
	"database/sql"
	"fmt"
	"net/http"
	"os/exec"
	"crypto/md5"
	"crypto/sha1"
	"crypto/des"
	"log"
	"io/ioutil"
	"path/filepath"
	_ "github.com/go-sql-driver/mysql"
)

// SQL Injection vulnerabilities
func vulnerableQuery(db *sql.DB, userID string) {
	// CWE-89: SQL Injection
	query := "SELECT * FROM users WHERE id = " + userID
	rows, err := db.Query(query)
	if err != nil {
		log.Fatal(err)
	}
	defer rows.Close()
	
	// Another SQL injection
	db.Exec("DELETE FROM users WHERE name = '" + userID + "'")
}

// Command Injection vulnerabilities
func vulnerableCommand(userInput string) {
	// CWE-78: Command Injection
	cmd := exec.Command("sh", "-c", "echo "+userInput)
	output, _ := cmd.Output()
	fmt.Println(string(output))
	
	// Another command injection
	exec.CommandContext(nil, "grep", userInput, "/etc/passwd")
}

// Path Traversal vulnerabilities
func vulnerableFileAccess(filename string) {
	// CWE-22: Path Traversal
	fullPath := filepath.Join("/var/data", filename)
	content, _ := ioutil.ReadFile(fullPath)
	fmt.Println(string(content))
}

// SSRF vulnerabilities
func vulnerableHTTPRequest(url string) {
	// CWE-918: Server-Side Request Forgery
	resp, err := http.Get(url)
	if err != nil {
		log.Fatal(err)
	}
	defer resp.Body.Close()
	
	// Another SSRF
	http.Post(url, "application/json", nil)
}

// Weak Cryptography
func weakCrypto() {
	// CWE-327: Weak cryptographic algorithms
	h1 := md5.New()
	h1.Write([]byte("password"))
	
	h2 := sha1.New()
	h2.Write([]byte("secret"))
	
	// Weak cipher
	des.NewCipher([]byte("weakkey1"))
}

// Hardcoded Secrets
func hardcodedCredentials() {
	// CWE-798: Hardcoded credentials
	password := "SuperSecret123!"
	apiKey := "sk_live_1234567890abcdef"
	dbConnection := "mysql://user:password@localhost/db"
	
	fmt.Println(password, apiKey, dbConnection)
}

// Race condition vulnerability
var counter int

func raceCondition() {
	// Potential race condition - no synchronization
	for i := 0; i < 1000; i++ {
		go func() {
			counter++ // Unsafe concurrent access
		}()
	}
}

// Memory leak potential
func memoryLeak() {
	// Potential goroutine leak
	for {
		go func() {
			// Goroutine that never exits
			select {}
		}()
	}
}

func main() {
	// Demonstrate vulnerabilities
	db, _ := sql.Open("mysql", "user:pass@/dbname")
	defer db.Close()
	
	vulnerableQuery(db, "1 OR 1=1")
	vulnerableCommand("$(rm -rf /)")
	vulnerableFileAccess("../../../etc/passwd")
	vulnerableHTTPRequest("http://internal.service/admin")
	weakCrypto()
	hardcodedCredentials()
	raceCondition()
	memoryLeak()
}